/**:
  ros__parameters:
    # -- Control Algorithm Parameters --
    control_frequency: 30.0          # Hertz
    mpc_prediction_dt: 0.1           # prediction time step length [s]
    cog_rear_lr: 0.0                 # distance of rear axle center to COG [m] -keep it zero

    # -- Vehicle Model Parameters --
    steering_time_constant: 0.24  # [s]
    speed_time_constant: 0.1    # [s]

    # -- Simulation model for planning simulator --
    use_mpc_controller: true    # if false it uses the feedback controllers.
    use_delay_sim_model: true   # for DELAY_STEER planning simulator model. For debugging only.
    input_delay_time: 0.24 # 0.3    # input dead-time [s]

    # -- Use time or arc-length as the interpolation coordinates for predictions --
    use_acceleration_inputs: true                 # if false, traj speeds are used in the simulations
    use_linear_trajectory_initialization: false    # if false LPV feedback initialization is used.

    # -- State and Control Weights.
    # -- ['xw', 'yw', 'psi', 's', 'e_y', 'e_yaw', 'Vx', 'delta', 'ay'] --
    # state_weights: [ 0.0, 0.0, 0.0, 0.0, 0.25, 0.25, 0.25, 0.0, 0.0 ]
    # state_weights_terminal: [ 0.0, 0.0, 0.0, 0.0, 0.25, 0.25, 0.25, 0.0, 0.0 ]

    state_weights: [ 0.0, 0.0, 0.0, 0.0, 0.01, 0.01, 0.5, 0.0, 0.0 ]
    state_weights_terminal: [ 0.0, 0.0, 0.0, 0.0, 0.02, 0.02, 0.5, 0.0, 0.0 ]
    control_weights: [ 0.1, 0.1 ]    # [vx speed input, steering input]
    jerk_weights: [ 0.5, 0.5 ]        # [acc and steering_rate]

    #  -- Constraints - Upper-lower bounds.
    #  -- ['xw', 'yw', 'psi', 's', 'e_y', 'e_yaw', 'Vx', 'delta', 'ay'] --
    # xlower: [ -.inf, -.inf, -.inf, -.inf, -3.0, -2.0, -0.5, -0.65, -.inf ]
    xlower: [ -.inf, -.inf, -.inf, -.inf, -1.0, -1.0, 0.0, -1.0, -4.0 ]
    xupper: [ .inf,  .inf,   .inf, .inf,  1.0,  1.0, 25.0, 1.0, 4.0 ]
    ulower: [ -7., -1.0 ]  # vx speed input [m/s], steering input [rad]
    uupper: [ 7.,  1.0 ]

    # -- Scaling Mats ['xw', 'yw', 'psi', 's', 'e_y', 'e_yaw', 'Vx', 'delta', 'ay'] --
    scaling_range: [ -1., 1. ]

    xmin_for_scaling: [ -25., -25., -1.0, 0.0,  -1.0, -1.0, 0.0, -1.0, -5.0 ]
    xmax_for_scaling: [ 25.,  25.,  1.0, 25.0,   1.0,  1.0, 10.0,  1.0, 5.0 ]

    umin_for_scaling: [ -7., -1.0 ]  # ax speed input [m/s], steering input [rad]
    umax_for_scaling: [ 7.,  1.0 ]

    # OSQP Options.
    osqp_warm_start: true
    osqp_polishing: true
    osqp_scaling: true
    osqp_max_iters: 500
    osqp_polish_iters: 5
    osqp_time_limit: 0.02 # seconds - currently disabled
    osqp_eps_abs: 0.0001
    osqp_eps_rel: 0.0001
    osqp_verbose: false     # outputs iteration results.
    osqp_scaled_termination: false

    # -- Lyapunov Matrices for LPV Initialization - Delay steering --
    # -- Error states are : [ey, eyaw, V, steering]
    # num_of_nonlinearities: 5  # Number of nonlinearities in the state transition matrix.

    Xn1: [ 0.00016595097571096674, -0.00015691798393613086, 0.00038907714650369155, 2.904603556981425e-07, -0.00015691798393613086, 8.572650658184816e-05, -7.738383536122305e-05, -1.4211604300928043e-07, 0.00038907714650369155, -7.738383536122305e-05, -0.0009012039764308131, -9.4070541504852e-07, 2.904603556981425e-07, -1.4211604300928043e-07, -9.4070541504852e-07, -4.4938947707522256e-08 ]
    Xn2: [ 0.01606329194398072, -0.011590957693481268, 0.011828962035058649, 1.7967130142460518e-05, -0.011590957693481268, 0.004872383779586471, -0.004361073211713394, -3.0082462718213966e-05, 0.011828962035058649, -0.004361073211713394, 0.004178257826895157, 3.556063921490683e-05, 1.7967130142460518e-05, -3.0082462718213966e-05, 3.556063921490683e-05, -5.049055510590558e-08 ]
    Xn3: [ 6.789299919275097e-05, -3.270051608674559e-05, 0.00017521016345111446, 7.285292914583504e-08, -3.270051608674559e-05, -0.0001377619027694065, 0.0006565854759100682, -4.754993099364052e-07, 0.00017521016345111446, 0.0006565854759100682, -0.0026540604958835163, 2.7927736836838043e-06, 7.285292914583504e-08, -4.754993099364052e-07, 2.7927736836838043e-06, 7.546105066199243e-08 ]
    Xn4: [ 4.502888366345244e-06, -7.350641754203101e-06, 3.097080918568864e-05, -6.549831719898833e-09, -7.350641754203101e-06, 1.3976461165308242e-05, -5.7921181415129755e-05, 5.720272106498826e-08, 3.097080918568864e-05, -5.7921181415129755e-05, 0.00017776548548544807, -4.2995672083163053e-07, -6.549831719898833e-09, 5.720272106498826e-08, -4.2995672083163053e-07, -1.7044147461008145e-08 ]
    Xn5: [ 0.10783587922129301, -0.02305981749020662, 0.004017084870545823, -7.791735555798708e-05, -0.02305981749020662, 0.014231403764046769, -0.013816712769741881, -4.1639359758227974e-05, 0.004017084870545823, -0.013816712769741881, 0.04201871371348142, 2.58815050968221e-05, -7.791735555798708e-05, -4.1639359758227974e-05, 2.58815050968221e-05, 0.05564551988431361 ]

    Yn1: [ -0.014932590556751632, 2.049565854453017e-06, 0.005085366699427884, 4.966019481536339e-06, -0.003542540570338895, -7.993527096450875e-06, -3.3203240899628235e-05, 3.0390224642778166e-09 ]
    Yn2: [ -0.07055723644803315, -1.6870752244453534e-06, -0.059599727759294674, 1.0806795965854139e-05, 0.0876446676281332, -1.742430609637487e-05, -0.00049316985507534, 6.599819612377416e-08 ]
    Yn3: [ -0.006092137171559172, 1.0336854843665861e-06, 0.0003543654402945735, -1.0514549283696234e-06, 0.002645202373731836, 1.82296609097731e-06, 2.0695294011271134e-05, 6.69779606466217e-09 ]
    Yn4: [ 0.025567720145854983, -2.0805278859698697e-06, -0.008731374747814234, -7.844507775047496e-06, 0.008652514443550781, 1.2492683820088609e-05, 4.538480849131615e-05, -2.2659987558210137e-09 ]
    Yn5: [ -0.0019637108188206138, 1.7457115568579532e-07, 0.00047218332019417574, 4.485751820151556e-07, -0.04760912772665273, -1.2441697430752518e-06, 1.0007116796972773e-05, -0.009823476533499251 ]

    # -- Stopping Condition Parameters --
    stop_state_entry_ego_speed: 0.2
    stop_state_entry_target_speed: 0.5
    stop_state_keep_stopping_dist: 0.5
    will_stop_state_dist: 1.0 # is used to get if vehicle is approaching to stopping point

    # -- Accuracy parameters --
    number_of_sqp_iterations: 1

    # CDOB - DOB parameters.
    use_cdob: true
    use_dob: true # activating use_dob only enables use_cdob. They are both used.

    # -- Kalman filter parameters for error model states [e_y, e_yaw[rad], v, steering [deg]] --
    # -- UKF specific parameters --
    kalman_filters:
      use_kalman: false
      use_sqrt_version: false
      alpha: 0.1  # controls spread of the sigma points,
      beta: 2.0   # for Gaussian distribution,
      kappa: 0.05  # scale the sigma points further from x_mean.

      #  -- ['xw', 'yw', 'psi', 's', 'e_y', 'e_yaw', 'Vx', 'delta', 'ay'] --
      Pkalman: [ 0.1, 0.1, 0.08, 0.3, 0.15, 0.07, 0.5, 0.07, 0.2 ]   # Kalman estimate uncertainty variance estimate.
      Vprocess: [ 0.2, 0.2, 0.05, 0.1, 0.02, 0.01, 0.2, 0.03, 0.1 ]  # process [model] uncertainty variance estimate.
      Wmeasurement: [ 0.4, 0.4, 0.06, 0.2,  0.15, 0.1, 0.1, 0.02, 0.2 ]  # measurement uncertainty variance estimate


    deadzone_params:
      use_deadzone_inv: true
      bmax: 0.06
      mr: 1.
      br: 0.05
      ml: 1.
      bl: 0.05

      use_extremum_seeker: false
      K_ex: 5. # Gain
      ay: 0.01 # dithering sinusoidal amplitude.
      freq_low_ex: 5. #low-pass filter cut-off frequency in Hertz
      freq_high_ex: 1. #high-pass filter cut-off frequency
      freq_dither_ex: 1.0 #dither signal frequency
